Break the Dedekind reals into lower and upper Dedekind reals (#1314)
UniMath/agda-unimath: src/real-numbers/dedekind-real-numbers.lagda.md

open import foundation-core.truncation-levels

open import logic.functoriality-existential-quantification

open import real-numbers.lower-dedekind-real-numbers
open import real-numbers.upper-dedekind-real-numbers
```

</details>

## Idea

> A Dedekind real number consists of a [pair](foundation.dependent-pair-types.md)
> `(lx , uy)` of a
> [lower Dedekind real](real-numbers.lower-dedekind-real-numbers.md) and an
> [upper Dedekind real](real-numbers.upper-dedekind-real-numbers.md) that also
> satisfy the following conditions:
> 
> 1. _Disjointness_. The cuts of `lx` and `uy` are disjoint subsets of `ℚ`.
> 2. _Locatedness_. If `q < r` then `q` is in the cut of `lx` or `r` is in the cut
>    of `uy`.

Reviewer: ```suggestion
A {{#concept "Dedekind real number" WD="real number" WDID=Q12916 Agda=ℝ}} consists of a [pair](foundation.dependent-pair-types.md)
 of _cuts_ `(lx , uy)` of [rational numbers](elementary-number-theory.rational-numbers.md), a
[lower Dedekind cut](real-numbers.lower-dedekind-real-numbers.md) `lx` and an
[upper Dedekind cut](real-numbers.upper-dedekind-real-numbers.md) `uy`. These cuts present lower and upper bounds on the Dedekind real number, and must satisfy the conditions

1. _Disjointness_. The cuts `lx` and `uy` are disjoint subsets of `ℚ`.
2. _Locatedness_. If `q < r` then `q` is in the cut `lx` or `r` is in the cut `uy`.

The {{#concept "collection of all Dedekind real numbers" WD="set of real numbers" WDID=Q1174982 Agda=ℝ}} is denoted `ℝ`.
```

>    of `uy`.

Reviewer: We should also explain what it means for a cut to be lower or upper

>    of `uy`.

Author: Do you have a preference for keeping the _implementation_ of a definition of a real number in terms of the two subtypes of the rationals, rather than the refactoring here?

>    of `uy`.

Author: Separately, I'm not sure what you have in mind with "explaining what it means for a cut to be upper or lower," when you've already added an inline link to what it means to be a lower Dedekind cut and an upper Dedekind cut.

>    of `uy`.

Reviewer: > Do you have a preference for keeping the _implementation_ of a definition of a real number in terms of the two subtypes of the rationals, rather than the refactoring here?

The implementation can use `lower-ℝ` and `upper-ℝ`, no problem. It is how we _interpret_ the elements of `lower-ℝ` and `upper-ℝ` that is the issue here.

>    of `uy`.

Reviewer: I suppose the current suggestion is not too bad in that it gives a half-explanation of what a lower and upper real number is, but given that `dedekind-real-numbers` is our most visited page, I think we should be careful to present it nicely

>    of `uy`.

Reviewer: More generally, it is an antipattern that we are all guilty of here to always link to every concept instead of explaining them. This gives an unpleasant experience to newcomers to the field when they have to jump from page to page just to get to a fundamental definition, from which they may have a long way back up to their originating point.

It is without question that "lower and upper Dedekind reals" are more niche concepts than "Dedekind reals" so we shouldn't expect a reader of this Idea section to know what they mean. Therefore we should explain it to them.

>    of `uy`.

Author: I've attempted to do that.

>    of `uy`.

Reviewer: Thank you for that. It looks better now

>    of `uy`.

Reviewer: To be more precise, the explanation is good enough for now. But we should rethink the exposition again later, given that this is, as I said, the most visited (formalization-)page of the library.

> 

Reviewer: To merge the above and below paragraph.

```suggestion
```

The Dedekind real numbers will be taken as the standard definition of the real
numbers in the agda-unimath library.

## Definition

> ### Dedekind real numbers

Reviewer: I think this section should still be phrased as being about cuts. Even though the "lower Dedekind reals" gives a half-model of real numbers, they're not used as an independent model of real numbers in the definition of Dedekind real numbers.

Reviewer: ```suggestion
### Dedekind cuts
```

          ( pr1 uy)
          ( pr1 uy')
          ( eq-upper-cut-eq-lower-cut-ℝ (lx , uy) (lx , uy') refl)))

is-emb-lower-real : {l : Level} → is-emb (lower-real-ℝ {l})
is-emb-lower-real = is-emb-inclusion-subtype has-upper-real-Prop
```

### The map from a real number to its upper real is an embedding

```agda
module _
  {l : Level}
  (uy : upper-ℝ l)
  where

  has-lower-real-Prop : Prop (lsuc l)
  pr1 has-lower-real-Prop =
    Σ (lower-ℝ l) (λ lx → is-dedekind-lower-upper-ℝ lx uy)
  pr2 has-lower-real-Prop =
    ( is-prop-all-elements-equal)
    ( λ lx lx' →
      eq-type-subtype
        ( λ l → is-dedekind-prop-lower-upper-ℝ l uy)
        ( eq-eq-cut-lower-ℝ
          ( pr1 lx)
          ( pr1 lx')
          ( eq-lower-cut-eq-upper-cut-ℝ
            ( pr1 lx , uy , pr2 lx)
            ( pr1 lx' , uy , pr2 lx')
            ( refl))))

is-emb-upper-real : {l : Level} → is-emb (upper-real-ℝ {l})
pr1 (pr1 (is-emb-upper-real (lx , ux , Hx) (ly , uy , Hy))) ux=uy =
  ap
    ( λ (uz , lz , Hz) → (lz , uz , Hz))
    ( pr1
      ( pr1
        ( is-emb-inclusion-subtype
          ( has-lower-real-Prop)
          ( ux , lx , Hx)
          ( uy , ly , Hy)))
      ( ux=uy))

> pr2 (pr1 (is-emb-upper-real (lx , ux , Hx) (ly , uy , Hy))) ux=uy =

Author: I'm struggling a bit here, but maybe the easiest path is to prove that reversing the role of the upper and lower cuts is equivalence, that that is an embedding, and then that the composition of two embeddings is an embedding...?

Author: Actually, realizing that the previous version of this file didn't prove this, I'm not going to worry too much.

          ( lower-cut-ℝ x)
          ( lower-cut-ℝ y)
          ( H)))

  eq-upper-real-eq-lower-real-ℝ :
    lower-real-ℝ x ＝ lower-real-ℝ y → upper-real-ℝ x ＝ upper-real-ℝ y
  eq-upper-real-eq-lower-real-ℝ lx=ly =
    eq-eq-cut-upper-ℝ
      ( upper-real-ℝ x)
      ( upper-real-ℝ y)
      ( eq-upper-cut-eq-lower-cut-ℝ (ap cut-lower-ℝ lx=ly))
```

### The map from a real number to its lower real is an embedding

```agda
module _
  {l : Level}
  (lx : lower-ℝ l)
  where

  has-upper-real-Prop : Prop (lsuc l)

>   pr1 has-upper-real-Prop = Σ (upper-ℝ l) (is-dedekind-lower-upper-ℝ lx)

Reviewer: From my point of view this should be phrased as a "`is-dedekind-cut-lower-ℝ-Prop`"
